### PR TITLE
Update fixing

### PR DESCRIPTION
**Pull request by afragen**
When I added another package ( did:plc:deoui6ztyx6paqajconl67rz ) for testing I found all sorts of edge cases for installing and updating from `plugins.php`, `core-update.php`, and from modals.

Pretty sure it's working now.

### CODE REVIEW

File: inc/updater/namespace.php [hunk 1]
```diff
@@ -25,9 +27,48 @@ function bootstrap() {
...
-	$metadata = $obj->metadata ?? $obj->package;
-	$release = $obj->release;
+	global  $release;
+
+	$packages = [];
+	// phpcs:disable HM.Security.NonceVerification.Recommended
+	if ( $obj instanceof Upgrader ) {
+		if ( isset( $_REQUEST['action'] ) && 'fair-install-plugin' === $_REQUEST['action'] ) {
+			$did = isset( $_REQUEST['id'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['id'] ) ) :
+			'';
```
Comment (costdev) [inline inc/updater/namespace.php]:
```suggestion
			$did = isset( $_REQUEST['id'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['id'] ) ) : '';
```

File: inc/updater/namespace.php [hunk 2]
```diff
@@ -55,19 +95,23 @@ function upgrader_pre_download() : bool {
...
+	foreach ( $release as $rel ) {
+		if ( $url === $rel->artifacts->package[0]->url ) {
+			foreach ( $rel->artifacts->package[0] as $key => $value ) {
+				$key = str_replace( '-', '_', $key );
+				$package[ $key ] = $value;
+			}
+			if ( isset( $package['content_type'] ) && $package['content_type'] === 'application/octet-stream' ) {
+				$accept_header = [ 'headers' => [ 'Accept' => 'application/octet-stream' ] ];
+				$args = array_merge( $args, $accept_header );
+			}
```
Comment (costdev) [inline inc/updater/namespace.php]: Is this replaceable if we use `$rel->artifacts->package[0]['content-type']`?

Something like this?

```php
$content_type = $rel->artifacts->package[0]['content-type'];
if ( $content_type === 'application/octet-stream' ) {
	$args = array_merge( $args, [ 'headers' => [ 'Accept' => $content_type ] ] );
}
```

File: inc/updater/namespace.php [hunk 3]
```diff
@@ -25,9 +27,47 @@ function bootstrap() {
...
+	$packages = [];
+	// phpcs:disable HM.Security.NonceVerification.Recommended
+	if ( $obj instanceof Upgrader ) {
+		if ( isset( $_REQUEST['action'] ) && 'fair-install-plugin' === $_REQUEST['action'] ) {
+			$did = isset( $_REQUEST['id'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['id'] ) ) : '';
+			if ( $did === $obj->package->id ) {
+				$release[ $did ] = $obj->release;
+			}
+		}
+	}
```
Comment (costdev) [inline inc/updater/namespace.php]: Optional: Could trim this down if we merge some conditions and save on entering the initial `if` at all.

```suggestion
	if (
		$obj instanceof Upgrader
		&& isset( $_REQUEST['action'], $_REQUEST['id'] )
		&& 'fair-install-plugin' === $_REQUEST['action']
	) {
		$did = sanitize_text_field( wp_unslash( $_REQUEST['id'] ) );
		if ( $did === $obj->package->id ) {
			$release[ $did ] = $obj->release;
		}
	}
```

**Review by costdev — APPROVED** [review]

**Comment by afragen** [comment]
Semi-large refactoring with some PHPCS ignoring when it makes sense.

File: inc/updater/namespace.php [hunk 4]
```diff
@@ -55,25 +104,55 @@ function upgrader_pre_download() : bool {
...
 
 	return $args;
 }
 
+/**
+ * Get data from DID.
+ *
+ * @param  string $id DID.
+ *
+ * @return mixed
```
Comment (costdev) [inline inc/updater/namespace.php]:
```suggestion
 * @return ReleaseDocument|WP_Error The latest release, or a WP_Error object on failure.
```

File: inc/updater/namespace.php [hunk 5]
```diff
@@ -55,25 +104,55 @@ function upgrader_pre_download() : bool {
...
+				$args = array_merge( $args, [ 'headers' => [ 'Accept' => $content_type ] ] );
+			}
 		}
 	}
 
 	return $args;
 }
 
+/**
+ * Get data from DID.
```
Comment (costdev) [inline inc/updater/namespace.php]:
```suggestion
 * Get the latest release for a DID.
```

File: inc/updater/namespace.php [hunk 6]
```diff
@@ -55,25 +104,55 @@ function upgrader_pre_download() : bool {
...
+ *
+ * @param  string $id DID.
+ *
+ * @return mixed
+ */
+function get_release_from_did( $id ) {
```
Comment (costdev) [inline inc/updater/namespace.php]:
```suggestion
function get_latest_release_from_did( $id ) {
```
Since this will only ever return the latest release, might as well be specific here.

Comment (afragen) [reply inc/updater/namespace.php]: You need to fix the call too.

File: inc/updater/class-updater.php [hunk 7]
```diff
@@ -116,7 +110,7 @@ public function run() {
 		if ( is_wp_error( $this->metadata ) ) {
 			return $this->metadata;
 		}
-		$this->release = $this->get_data_from_did( $this->did );
+		$this->release = get_release_from_did( $this->did );
```
Comment (costdev) [inline inc/updater/class-updater.php]:
```suggestion
		$this->release = get_latest_release_from_did( $this->did );
```

**Review by costdev — APPROVED** [review]
